rust/gateway/src/eventloop.rs

```diff
@@ -0,0 +1,193 @@
+use crate::control::ControlSignaler;
+use crate::messages::{
+    AllowAccess, BroadcastClientIceCandidates, ClientIceCandidates, ConnectionReady,
+    EgressMessages, IngressMessages,
+};
+use crate::CallbackHandler;
+use anyhow::Result;
+use connlib_shared::messages::ClientId;
+use connlib_shared::{Callbacks, Error};
+use firezone_tunnel::Tunnel;
+use phoenix_channel::PhoenixChannel;
+use std::convert::Infallible;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use std::time::Duration;
+use tokio::sync::mpsc;
+use webrtc::peer_connection::sdp::session_description::RTCSessionDescription;
+
+pub struct Eventloop {
+    tunnel: Arc<Tunnel<ControlSignaler, CallbackHandler>>,
+    control_rx: mpsc::Receiver<BroadcastClientIceCandidates>,
+    portal: PhoenixChannel<IngressMessages, ()>,
+
+    // TODO: Strongly type request reference (currently `String`)
+    connection_request_tasks:
+        futures_bounded::FuturesMap<(ClientId, String), Result<RTCSessionDescription, Error>>,
+    add_ice_candidate_tasks: futures_bounded::FuturesSet<Result<(), Error>>,
+
+    print_stats_timer: tokio::time::Interval,
+}
+
+impl Eventloop {
+    pub(crate) fn new(
+        tunnel: Arc<Tunnel<ControlSignaler, CallbackHandler>>,
+        control_rx: mpsc::Receiver<BroadcastClientIceCandidates>,
+        portal: PhoenixChannel<IngressMessages, ()>,
+    ) -> Self {
+        Self {
+            tunnel,
+            control_rx,
+            portal,
+
+            // TODO: Pick sane values for timeouts and size.
+            connection_request_tasks: futures_bounded::FuturesMap::new(
+                Duration::from_secs(60),
+                100,
+            ),
+            add_ice_candidate_tasks: futures_bounded::FuturesSet::new(Duration::from_secs(60), 100),
+            print_stats_timer: tokio::time::interval(Duration::from_secs(10)),
+        }
+    }
+}
+
+impl Eventloop {
+    pub fn poll(&mut self, cx: &mut Context<'_>) -> Poll<Result<Infallible>> {
+        loop {
+            if let Poll::Ready(Some(ice_candidates)) = self.control_rx.poll_recv(cx) {
+                let _id = self.portal.send(
+                    "gateway",
```

Comment:
I am gonna move it to a const for now. It will be resolved properly once we have https://github.com/firezone/firezone/pull/2244#discussion_r1346818827.